File: src/lib/events.ts
// Tauri event names emitted by the Rust backend. Centralized so the (high
// frequency) mic-level event can be subscribed to by a single leaf component.

export const EVT = {
  /** ~50-100ms cadence while recording: { db, peak } in 0..1. High frequency. */
  micLevel: "voxctl:mic-level",
  /** Recording lifecycle: { recording: boolean, language: string | null }. */
  recState: "voxctl:rec-state",
  /** Final transcript on stop: { text }. */
  transcriptFinal: "voxctl:transcript-final",
  /** Active mode changed (auto-switch or manual): { id, name }. */
  modeChanged: "voxctl:mode-changed",
  /** History table changed; UI should refetch: {}. */
  historyChanged: "voxctl:history-changed",
  /** User changed language from the HUD picker: { language: string | null }. */
  langChanged: "voxctl:lang-changed",
  /** Non-fatal backend error surfaced to the UI: { stage, message }. */
  error: "voxctl:error",
} as const;

export interface MicLevel {
  db: number;
  peak: number;
}
export interface RecState {
  recording: boolean;
  language: string | null;
}
export interface ModeChanged {
  id: string;
  name: string;
}
export interface BackendError {
  stage: string;
  message: string;
}

File: src-tauri/src/events.rs
//! Event names + payloads emitted to the webview. Mirrors `src/lib/events.ts`.
//! Some names/payloads are consumed in later slices.
#![allow(dead_code)]

use serde::Serialize;

pub const MIC_LEVEL: &str = "voxctl:mic-level";
pub const REC_STATE: &str = "voxctl:rec-state";
pub const TRANSCRIPT_FINAL: &str = "voxctl:transcript-final";
pub const MODE_CHANGED: &str = "voxctl:mode-changed";
pub const HISTORY_CHANGED: &str = "voxctl:history-changed";
pub const ERROR: &str = "voxctl:error";

#[derive(Clone, Serialize)]
pub struct MicLevel {
    /// Normalized 0..1 level (from RMS dBFS) for the meters.
    pub db: f32,
    pub peak: f32,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecState {
    pub recording: bool,
    pub language: Option<String>,
}

#[derive(Clone, Serialize)]
pub struct TranscriptText {
    pub text: String,
}

#[derive(Clone, Serialize)]
pub struct ModeChanged {
    pub id: String,
    pub name: String,
}

#[derive(Clone, Serialize)]
pub struct BackendError {
    pub stage: String,
    pub message: String,
}

#[allow(dead_code)]
impl BackendError {
    pub fn new(stage: &str, message: impl Into<String>) -> Self {
        Self {
            stage: stage.into(),
            message: message.into(),
        }
    }
}
